extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeSearchExtractor.java

```diff
@@ -49,52 +71,169 @@ public YoutubeSearchExtractor(StreamingService service, SearchQueryHandler linkH
 
-        final JsonArray ajaxJson = getJsonResponse(url, getExtractorLocalization());
+            final String url = "https://music.youtube.com/youtubei/v1/search?alt=json&key=" + youtubeMusicKeys[0];
 
-        initialData = ajaxJson.getObject(1).getObject("response");
+            String params = null;
+
+            switch (getLinkHandler().getContentFilters().get(0)) {
+                case MUSIC_SONGS:
+                    params = "Eg-KAQwIARAAGAAgACgAMABqChAEEAUQAxAKEAk%3D";
+                    break;
+                case MUSIC_VIDEOS:
+                    params = "Eg-KAQwIABABGAAgACgAMABqChAEEAUQAxAKEAk%3D";
+                    break;
+                case MUSIC_ALBUMS:
+                    params = "Eg-KAQwIABAAGAEgACgAMABqChAEEAUQAxAKEAk%3D";
+                    break;
+                case MUSIC_PLAYLISTS:
+                    params = "Eg-KAQwIABAAGAAgACgBMABqChAEEAUQAxAKEAk%3D";
+                    break;
+                case MUSIC_ARTISTS:
+                    params = "Eg-KAQwIABAAGAAgASgAMABqChAEEAUQAxAKEAk%3D";
+                    break;
+            }
+
+            // @formatter:off
+            byte[] json = JsonWriter.string()
+                .object()
+                    .object("context")
+                        .object("client")
+                            .value("clientName", "WEB_REMIX")
+                            .value("clientVersion", youtubeMusicKeys[2])
+                            .value("hl", "en")
+                            .value("gl", getExtractorContentCountry().getCountryCode())
+                            .array("experimentIds").end()
+                            .value("experimentsToken", "")
+                            .value("utcOffsetMinutes", 0)
+                            .object("locationInfo").end()
+                            .object("musicAppInfo").end()
+                        .end()
+                        .object("capabilities").end()
+                        .object("request")
+                            .array("internalExperimentFlags").end()
+                            .object("sessionIndex").end()
+                        .end()
+                        .object("activePlayers").end()
+                        .object("user")
+                            .value("enableSafetyMode", false)
+                        .end()
+                    .end()
+                    .value("query", getSearchString())
+                    .value("params", params)
+                .end().done().getBytes("UTF-8");
+            // @formatter:on
+
+            Map<String, List<String>> headers = new HashMap<>();
+            headers.put("X-YouTube-Client-Name", Collections.singletonList(youtubeMusicKeys[1]));
+            headers.put("X-YouTube-Client-Version", Collections.singletonList(youtubeMusicKeys[2]));
+            headers.put("Origin", Collections.singletonList("https://music.youtube.com"));
```

Comment:
As discussed, the default http client from java [has some restrictions about request headers](https://hg.openjdk.java.net/jdk8/jdk8/jdk/file/687fd7c7986d/src/share/classes/sun/net/www/protocol/http/HttpURLConnection.java#l194).

I think using another client for tests is the right choice (like you did earlier, OkHttp is a good one).

Should we do it in its own pull request? The front end is already using it, don't think too many problems would be a result of this, but then we can test thoroughly what kind of effects it will have.

Reply:
This should indeed be a different PR. I think OkHttp should be used, so we have the exact same behavior in the tests as in NewPipe.